File: server/src/inku_server/render_engines/default/determinism.py
"""Deterministic seed, hash, and value-noise primitives for the default engine."""

from __future__ import annotations

import hashlib
import json
import math
import secrets
import struct

from ...schema import Instruction, Variation

_VARIATION_SEED_FIELDS_ALL = frozenset(
    {"amplitude", "frequency", "quality", "dimensions"}
)
_WORK_COLOR_SEED_FIELDS = ("render_seed", "catalog_id", "abstract_color")

_SEED_INSTRUCTION_FIELDS = (
    "primitive",
    "from_",
    "to",
    "center",
    "radius",
    "sides",
    "position",
    "size",
    "angle_start",
    "angle_end",
    "rotation",
    "filled",
    "style",
    "weight",
    "thinness",
    "mode",
    "carve_depth",
    "variation",
    "arrangement",
    "surface",
)
_SEED_ARRANGEMENT_FIELDS = ("jitter",)


def _variation_seed_fields(ins: Instruction) -> frozenset[str] | None:
    """seed key に残す variation フィールド。None = variation ごと落とす。

    演奏されない variation が seed を動かすと、同じ意図が weight 次第で別の
    演奏になる。実際に消費されるフィールドだけを残す。判定は primitive ごとに
    異なる (cloudform は dimensions を見ず、残り 3 つを輪郭生成器の引数に使う)。
    """
    variation = ins.variation
    if variation is None:
        return None
    if ins.primitive == "cloudform":
        return frozenset({"amplitude", "frequency", "quality"})
    if _needs_blur(variation):
        # 滲みは stdDeviation だけを代表寸法と amplitude から決める。
        return frozenset({"amplitude", "quality"})
    if ins.primitive == "line":
        return _VARIATION_SEED_FIELDS_ALL if _needs_path_variation(variation) else None
    if _needs_contour_variation(variation):
        return _VARIATION_SEED_FIELDS_ALL
    return None


def _seed_for_instruction(ins: Instruction, performance_seed: int | None = None) -> int:
    """Instruction と演奏 seed から安定した乱数 seed を作る。"""
    dumped = ins.model_dump(mode="json")
    payload = {name: dumped[name] for name in _SEED_INSTRUCTION_FIELDS}
    arrangement = payload.get("arrangement")
    if isinstance(arrangement, dict):
        payload["arrangement"] = {
            name: arrangement[name] for name in _SEED_ARRANGEMENT_FIELDS
        }
    if payload.get("mode") == "additive":
        payload.pop("mode", None)
    if payload.get("carve_depth") is None:
        payload.pop("carve_depth", None)
    variation_payload = payload.get("variation")
    if isinstance(variation_payload, dict):
        fields = _variation_seed_fields(ins)
        if fields is None:
            # variation なしの Instruction と同じ key にする (pop ではなく None)。
            payload["variation"] = None
        elif fields is not _VARIATION_SEED_FIELDS_ALL:
            payload["variation"] = {
                name: value
                for name, value in variation_payload.items()
                if name in fields
            }
    surface = payload.get("surface")
    if isinstance(surface, dict) and surface.get("texture") == "solid":
        # 演奏されないフィールドを key から外す作法 (`_variation_seed_fields` と同じ)。
        # `solid` は素材の既定の埋め方で、面の質感の層はそれを 1 本も引かないから、
        # 残りの surface フィールドは何も消費されない。`filled=true` と書いた
        # Instruction と同じ key にする — 同じ要求の 2 通りの言い方が別の演奏に
        # なってはいけない。これが無いと、coerce が保存済みの `filled` へ
        # `surface` を足した瞬間に本番の全作品の筆致が振り直される。
        payload["filled"] = True
        payload["surface"] = surface = None
    if isinstance(surface, dict):
        if surface.get("spacing_gradient") == "none":
            surface.pop("spacing_gradient", None)
        if surface.get("tone_steps") == 3:
            surface.pop("tone_steps", None)
    key = json.dumps(payload, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    if performance_seed is not None:
        key += f":render:{performance_seed}".encode("utf-8")
    digest = hashlib.sha256(key).digest()
    return struct.unpack("<Q", digest[:8])[0]


def new_render_seed() -> int:
    """演奏ごとのマクロ揺らぎ seed。明示 seed 指定時は再現可能。"""
    return secrets.randbits(53)


def _hash_to_unit(i: int, seed: int) -> float:
    """i と seed から [-1, 1] の擬似乱数値を決定的に生成。"""
    h = hashlib.sha256(f"{seed}:{i}".encode("utf-8")).digest()
    val = struct.unpack("<q", h[:8])[0]
    return val / float(2**63)


def _value_noise_1d(x: float, seed: int) -> float:
    """smoothstep 補間の 1D value noise (擬似 perlin)。"""
    xi = math.floor(x)
    xf = x - xi
    v1 = _hash_to_unit(xi, seed)
    v2 = _hash_to_unit(xi + 1, seed)
    t = xf * xf * (3 - 2 * xf)
    return v1 * (1 - t) + v2 * t


def _needs_blur(v: Variation | None) -> bool:
    """quality=pink → SVG feGaussianBlur で滲み表現。"""
    return v is not None and v.quality == "pink"


def _needs_path_variation(v: Variation | None) -> bool:
    """quality=perlin/wave/white かつ dimensions 指定あり → polyline 揺らぎ。pink は blur で処理。"""
    if v is None:
        return False
    if v.quality in ("none", "pink"):
        return False
    return any(d in ("position_x", "position_y") for d in v.dimensions)


_CONTOUR_VARIATION_DIMS = ("position_x", "position_y", "radius")


def _needs_contour_variation(v: Variation | None) -> bool:
    """弧・閉図形の輪郭揺らぎ。line のゲートと対称で、図形の自然軸として radius を加える。"""
    if v is None:
        return False
    if v.quality in ("none", "pink"):
        return False
    return any(d in _CONTOUR_VARIATION_DIMS for d in v.dimensions)


def _wave_phase(seed: int) -> float:
    """seed から [0, 2π) の位相を決定的に導出する。

    wave は sin の位相が固定だと演奏 seed に依存せず、同じ Score が常に同じ
    山谷の位置になる。位相だけを seed 由来にすることで、周期性 (整数周波数
    なら t∈[0,1) で閉じる) と振幅・周波数の語彙を保ったまま演奏差を作る。
    """
    return _hash01(0, seed, "wave-phase") * 2 * math.pi


def _periodic_value_noise_1d(x: float, seed: int, period: int) -> float:
    """周期境界つき value noise。閉じた輪郭の継ぎ目を連続にする。"""
    xi = math.floor(x)
    xf = x - xi
    v1 = _hash_to_unit(int(xi) % period, seed)
    v2 = _hash_to_unit((int(xi) + 1) % period, seed)
    t = xf * xf * (3 - 2 * xf)
    return v1 * (1 - t) + v2 * t


def _hash01(i: int, seed: int, salt: str) -> float:
    h = hashlib.sha256(f"{seed}:{salt}:{i}".encode()).digest()
    return struct.unpack("<I", h[:4])[0] / 0xFFFFFFFF

File: core/crates/inku-render/src/determinism.rs
//! Deterministic hashes, seed payloads, and scalar noise.

use serde::Serialize;
use sha2::{Digest, Sha256};

use crate::types::{
    Amplitude, Arrangement, CarveDepth, Dimension, Frequency, Instruction, InstructionMode,
    LineStyle, Point, Primitive, Quality, Seed, SurfaceDirection, SurfaceSpacingGradient,
    SurfaceSpec, SurfaceTexture, Thinness, Variation, Weight,
};

fn sha256(payload: &[u8]) -> [u8; 32] {
    Sha256::digest(payload).into()
}

/// The `Display` text of one integer, written into a stack buffer.
///
/// Seeded hashes are defined over decimal text. Writing the digits directly
/// avoids a heap string per sample while hashing exactly the same bytes.
struct DecimalText {
    // 39 digits of u128::MAX plus a sign.
    buffer: [u8; 40],
    start: usize,
}

impl DecimalText {
    fn new(value: i128) -> Self {
        let mut buffer = [0_u8; 40];
        let mut start = buffer.len();
        let magnitude = value.unsigned_abs();
        // Seeds are usually u64-derived; u64 division is much cheaper than u128.
        if let Ok(mut rest) = u64::try_from(magnitude) {
            loop {
                start -= 1;
                buffer[start] = b'0' + (rest % 10) as u8;
                rest /= 10;
                if rest == 0 {
                    break;
                }
            }
        } else {
            let mut rest = magnitude;
            loop {
                start -= 1;
                buffer[start] = b'0' + (rest % 10) as u8;
                rest /= 10;
                if rest == 0 {
                    break;
                }
            }
        }
        if value < 0 {
            start -= 1;
            buffer[start] = b'-';
        }
        Self { buffer, start }
    }

    fn as_bytes(&self) -> &[u8] {
        &self.buffer[self.start..]
    }
}

/// SHA-256 of the text `"{seed}:{index}"`.
fn seed_index_digest(seed: Seed, index: i128) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(DecimalText::new(seed).as_bytes());
    hasher.update(b":");
    hasher.update(DecimalText::new(index).as_bytes());
    hasher.finalize().into()
}

/// SHA-256 of the text `"{seed}:{salt}:{index}"`.
///
/// The pieces are streamed in order, so the digest equals the digest of the
/// formatted string. Every seeded stream keeps its identity; only the
/// per-call string allocation of the hot stroke and noise paths is gone.
pub(crate) fn seed_salt_index_digest(seed: Seed, salt: &str, index: i128) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(DecimalText::new(seed).as_bytes());
    hasher.update(b":");
    hasher.update(salt.as_bytes());
    hasher.update(b":");
    hasher.update(DecimalText::new(index).as_bytes());
    hasher.finalize().into()
}

/// Map an integer and seed to the same signed unit interval used by Engine 40.
#[must_use]
pub fn hash_to_unit(index: i64, seed: Seed) -> f64 {
    let digest = seed_index_digest(seed, i128::from(index));
    let value = i64::from_le_bytes(digest[..8].try_into().expect("eight digest bytes"));
    value as f64 / 2_f64.powi(63)
}

/// Map a salted integer and seed to the closed interval from zero to one.
#[must_use]
pub fn hash01(index: i64, seed: Seed, salt: &str) -> f64 {
    let digest = seed_salt_index_digest(seed, salt, i128::from(index));
    let value = u32::from_le_bytes(digest[..4].try_into().expect("four digest bytes"));
    f64::from(value) / f64::from(u32::MAX)
}

/// Smoothly interpolate deterministic lattice values around `x`.
#[must_use]
pub fn value_noise_1d(x: f64, seed: Seed) -> f64 {
    let xi = x.floor();
    let xf = x - xi;
    let lower = hash_to_unit(xi as i64, seed);
    let upper = hash_to_unit(xi as i64 + 1, seed);
    let smooth = xf * xf * (3.0 - 2.0 * xf);
    lower * (1.0 - smooth) + upper * smooth
}

/// Value noise whose lattice closes after `period` cells.
#[must_use]
pub fn periodic_value_noise_1d(x: f64, seed: Seed, period: i64) -> f64 {
    assert!(period > 0, "periodic noise requires a positive period");
    let xi = x.floor() as i64;
    let xf = x - xi as f64;
    let lower = hash_to_unit(xi.rem_euclid(period), seed);
    let upper = hash_to_unit((xi + 1).rem_euclid(period), seed);
    let smooth = xf * xf * (3.0 - 2.0 * xf);
    lower * (1.0 - smooth) + upper * smooth
}

/// Seed-derived phase that preserves an integer-frequency periodic seam.
#[must_use]
pub fn wave_phase(seed: Seed) -> f64 {
    hash01(0, seed, "wave-phase") * std::f64::consts::TAU
}

#[derive(Clone, Copy)]
enum VariationSeedFields {
    None,
    AmplitudeQuality,
    AmplitudeFrequencyQuality,
    All,
}

#[must_use]
pub fn needs_blur(variation: &Variation) -> bool {
    variation.quality == Quality::Pink
}

#[must_use]
pub fn needs_path_variation(variation: &Variation) -> bool {
    !matches!(variation.quality, Quality::None | Quality::Pink)
        && variation
            .dimensions
            .iter()
            .any(|dimension| matches!(dimension, Dimension::PositionX | Dimension::PositionY))
}

#[must_use]
pub fn needs_contour_variation(variation: &Variation) -> bool {
    !matches!(variation.quality, Quality::None | Quality::Pink)
        && variation.dimensions.iter().any(|dimension| {
            matches!(
                dimension,
                Dimension::PositionX | Dimension::PositionY | Dimension::Radius
            )
        })
}

fn variation_seed_fields(instruction: &Instruction) -> VariationSeedFields {
    let Some(variation) = instruction.variation.as_ref() else {
        return VariationSeedFields::None;
    };
    if instruction.primitive == Primitive::Cloudform {
        return VariationSeedFields::AmplitudeFrequencyQuality;
    }
    if needs_blur(variation) {
        return VariationSeedFields::AmplitudeQuality;
    }
    if instruction.primitive == Primitive::Line {
        return if needs_path_variation(variation) {
            VariationSeedFields::All
        } else {
            VariationSeedFields::None
        };
    }
    if needs_contour_variation(variation) {
        VariationSeedFields::All
    } else {
        VariationSeedFields::None
    }
}

#[derive(Serialize)]
struct SeedVariation {
    #[serde(skip_serializing_if = "Option::is_none")]
    amplitude: Option<Amplitude>,
    #[serde(skip_serializing_if = "Option::is_none")]
    frequency: Option<Frequency>,
    #[serde(skip_serializing_if = "Option::is_none")]
    quality: Option<Quality>,
    #[serde(skip_serializing_if = "Option::is_none")]
    dimensions: Option<Vec<Dimension>>,
}

impl SeedVariation {
    fn filtered(variation: &Variation, fields: VariationSeedFields) -> Option<Self> {
        match fields {
            VariationSeedFields::None => None,
            VariationSeedFields::AmplitudeQuality => Some(Self {
                amplitude: Some(variation.amplitude),
                frequency: None,
                quality: Some(variation.quality),
                dimensions: None,
            }),
            VariationSeedFields::AmplitudeFrequencyQuality => Some(Self {
                amplitude: Some(variation.amplitude),
                frequency: Some(variation.frequency),
                quality: Some(variation.quality),
                dimensions: None,
            }),
            VariationSeedFields::All => Some(Self {
                amplitude: Some(variation.amplitude),
                frequency: Some(variation.frequency),
                quality: Some(variation.quality),
                dimensions: Some(variation.dimensions.clone()),
            }),
        }
    }
}

#[derive(Serialize)]
struct SeedArrangement {
    jitter: f64,
}

impl From<&Arrangement> for SeedArrangement {
    fn from(value: &Arrangement) -> Self {
        Self {
            jitter: value.jitter,
        }
    }
}

#[derive(Serialize)]
struct SeedSurface {
    texture: SurfaceTexture,
    density: f64,
    scale: f64,
    opacity: f64,
    bleed: f64,
    direction: SurfaceDirection,
    #[serde(skip_serializing_if = "Option::is_none")]
    spacing_gradient: Option<SurfaceSpacingGradient>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tone_steps: Option<u8>,
    seed: Option<Seed>,
}

impl From<&SurfaceSpec> for SeedSurface {
    fn from(value: &SurfaceSpec) -> Self {
        Self {
            texture: value.texture,
            density: value.density,
            scale: value.scale,
            opacity: value.opacity,
            bleed: value.bleed,
            direction: value.direction,
            spacing_gradient: (value.spacing_gradient != SurfaceSpacingGradient::None)
                .then_some(value.spacing_gradient),
            tone_steps: (value.tone_steps != 3).then_some(value.tone_steps),
            seed: value.seed,
        }
    }
}

#[derive(Serialize)]
struct SeedPayload {
    // Surface intensity changes deposition only, never geometry or the seeded field placement.
    primitive: Primitive,
    from_: Option<Point>,
    to: Option<Point>,
    center: Option<Point>,
    radius: Option<f64>,
    sides: Option<u8>,
    position: Option<Point>,
    size: Option<Point>,
    angle_start: Option<f64>,
    angle_end: Option<f64>,
    rotation: Option<f64>,
    filled: bool,
    style: LineStyle,
    weight: Weight,
    thinness: Option<Thinness>,
    #[serde(skip_serializing_if = "Option::is_none")]
    mode: Option<InstructionMode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    carve_depth: Option<CarveDepth>,
    variation: Option<SeedVariation>,
    arrangement: Option<SeedArrangement>,
    surface: Option<SeedSurface>,
}

impl From<&Instruction> for SeedPayload {
    fn from(instruction: &Instruction) -> Self {
        let solid = instruction
            .surface
            .as_ref()
            .is_some_and(|surface| surface.texture == SurfaceTexture::Solid);
        Self {
            primitive: instruction.primitive,
            from_: instruction.from_,
            to: instruction.to,
            center: instruction.center,
            radius: instruction.radius,
            sides: instruction.sides,
            position: instruction.position,
            size: instruction.size,
            angle_start: instruction.angle_start,
            angle_end: instruction.angle_end,
            rotation: instruction.rotation,
            filled: instruction.filled || solid,
            style: instruction.style,
            weight: instruction.weight,
            thinness: instruction.thinness,
            mode: (instruction.mode_ != InstructionMode::Additive).then_some(instruction.mode_),
            carve_depth: instruction.carve_depth,
            variation: instruction.variation.as_ref().and_then(|variation| {
                SeedVariation::filtered(variation, variation_seed_fields(instruction))
            }),
            arrangement: instruction.arrangement.as_ref().map(SeedArrangement::from),
            surface: if solid {
                None
            } else {
                instruction.surface.as_ref().map(SeedSurface::from)
            },
        }
    }
}

/// Derive the stable instruction seed from Engine 40's canonical payload.
#[must_use]
pub fn instruction_seed(instruction: &Instruction, performance_seed: Option<Seed>) -> Seed {
    let mut key = serde_json::to_vec(&SeedPayload::from(instruction))
        .expect("seed payload contains only serializable canonical values");
    if let Some(seed) = performance_seed {
        key.extend_from_slice(format!(":render:{seed}").as_bytes());
    }
    let digest = sha256(&key);
    i128::from(u64::from_le_bytes(
        digest[..8].try_into().expect("eight digest bytes"),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn streamed_seed_digest_hashes_the_formatted_text() {
        let values = [
            0,
            7,
            -1,
            i128::from(i64::MIN),
            i128::from(u64::MAX),
            i128::from(u64::MAX) + 1,
            i128::MIN,
            i128::MAX,
        ];
        for seed in values {
            for index in values {
                assert_eq!(
                    seed_salt_index_digest(seed, "salt", index),
                    sha256(format!("{seed}:salt:{index}").as_bytes())
                );
                assert_eq!(
                    seed_index_digest(seed, index),
                    sha256(format!("{seed}:{index}").as_bytes())
                );
            }
        }
    }
}
